Adding global tree of bounding boxes

GivAlz opened the pull request:
Hi everyone,
  while this PR is not complete (the test is missing, but the code should be ok), I have a few questions thus I think it's better to post already.

Missing:
- Test (I only have to properly add it)
- Improve documentation

Questions:
- Currently the new function is in grid_tools, any better recommendation? @luca-heltai you were proposing numerics, but where exactly?
- The template of the tree takes also the number of leafs. Now, clearly we can't initialize this for every value from 1 to infinity (or a very big number). What about some powers of 2: for instance, 2, 4, 8, 16, 32, 64? (currently I'm initializing only a value).
- To avoid horribly long lines I'm using some boost namespaces. I am quite sure adding them in the middle of grid tools is not the best thing, but I couldn't understand what's the best way to do it.
- There are different kinds of trees implemented in boost, should I implement different versions of the function?

Thank you @Rombur for your help last week!!!
Best,
Giovanni

Rombur commented on source/grid/grid_tools.cc:
```diff
@@ -4892,6 +4892,65 @@ namespace GridTools
+
+    // Building a tree out of the bounding boxes
+    // We avoid using the insert method so that boost uses the packing algorithm
+    LinearPointRtree<spacedim, leaf_number> rstar_tree(
```
I don't think that you are building an R*-tree so the name is strange.

luca-heltai commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -2651,6 +2654,39 @@ namespace GridTools
+  template <int spacedim, int leaf_number>
+  using LinearPointRtree = bgi::rtree<std::tuple<bgBox<spacedim>, unsigned int>,
+                                      bgi::linear<leaf_number>>;
+
```
This should really be `using LinearBoxRtree = bgi::rtree<BoundingBox<spacedim>, bgi::linear<leaf_number>>`. In this way the rtree algorithm can use anything that has a `BoundingBox<spacedim>` as first element of the tuple/pair, or just a box. 

Notice the name. You copied this from the test where I was creating a tree of points (hence that name: `LinearPointRTree`), but you want t tree of boxes...

luca-heltai replied:
From boost documentation:


> The default IndexableGetter can translate all types adapted to Point, Box or Segment concepts (called Indexables). Furthermore, it can handle std::pair<Indexable, T>, boost::tuple<Indexable, ...> and std::tuple<Indexable, ...> when possible. For example, for Value of type std::pair<Box, int>, the default IndexableGetter translates from std::pair<Box, int> const& to Box const&.

luca-heltai replied:
if you define the rtree with indexable `BoundingBox`, your code further down will still work.

luca-heltai commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -2651,6 +2654,39 @@ namespace GridTools
+   * Given a vector of bounding boxes describing the portion of the
+   * triangulation covered by locally owned cells and the relative
+   * parallel::distributed::Triangulation , compute the
+   * linear r-tree, used for search.
```
I think if you add boxes *after* building the tree with a range, then the linear algorithm is used.

luca-heltai commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -2651,6 +2654,39 @@ namespace GridTools
+   */
+  template <int dim, int spacedim, int leaf_number>
+  std::tuple<LinearPointRtree<spacedim, leaf_number>,
+             std::vector<std::tuple<bgBox<spacedim>, unsigned int>>>
```
same here. The tree should be templated on the box type, not on derived tuples/pairs/etc of the box.

GivAlz commented:
So I used the name "RTree" in the end (maybe RTreePacking would be better?)...since the packing algorithm is much better then the others I wouldn't bother with other kind of trees.

For the same reason I hardcoded a fixed number of leafs (16) and removed the template.

I removed the spaces used for boost, as we're really using only one function from it (once we're using the dealii bounding boxes, I changed this too): I hope the way I did it is ok.

masterleinad commented:
@luca-heltai Have your requested changes been addressed?

masterleinad commented:
/run-tests

luca-heltai commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -2651,6 +2654,35 @@ namespace GridTools
+  template <int spacedim>
+  using Rtree = boost::geometry::index::rtree<
+    std::tuple<BoundingBox<spacedim>, unsigned int>,
+    boost::geometry::index::linear<16>>;
```
I've looked at this patch again, and I'm not sure I'm in favour of adding this global `Rtree` declaration. `Rtree` is a general name, but this is a very specific type, only used in one function, and most likely not useful for other purposes where and `Rtree` would also be useful. 

Maybe it would make sense to expose a more general `Rtree` alias (and *document* it.... there is no documentation whatsoever of what `Rtree` is here, what algorithm uses, etc.). And with the correct `CamelCase` name: `RTree`, i.e., 
~~~
template <typename LeafType>
  using RTree = boost::geometry::index::rtree<
    LeafType,
    boost::geometry::index::linear<16>>;
~~~
and provide a function like
~~~
template<typename LeafTypeIterator> 
Rtree<typename LeafTypeIterator::value_type>
pack_rtree(LeafTypeIterator begin, LeafTypeIterator end);
~~~
to automatically generate the right `RTree`.

luca-heltai commented:
@GivAlz , can you rebase these changes to use the existing `RTree` alias in `deal.II/numerics/rtree.h`?

GivAlz commented:
So I've almost finished the modifications @luca-heltai  asked (it really took long); but I have a problem:
deal.II compiles just fine, but if I try to compile the test I'll get the following error when calling the function:

> error: undefined reference to 'boost::geometry::index::rtree<std::pair<dealii::BoundingBox<2, double>, unsigned int>, boost::geometry::index::linear<16ul, 4ul>, boost::geometry::index::indexable<std::pair<dealii::BoundingBox<2, double>, unsigned int> >, boost::geometry::index::equal_to<std::pair<dealii::BoundingBox<2, double>, unsigned int> >, boost::container::new_allocator<std::pair<dealii::BoundingBox<2, double>, unsigned int> > > dealii::GridTools::build_global_description_tree<2, 2>(std::vector<dealii::BoundingBox<2, double>, std::allocator<dealii::BoundingBox<2, double> > > const&, dealii::parallel::distributed::Triangulation<2, 2> const&)'

Luca, do you know what I am missing? I tried to add all the possible boost.h files and copied your construction of RTree using a pair (modifying it to bounding boxes), but it seems boost doesn't find the right function...

GivAlz commented:
Ok, I've found the problem (it was me, not initializing everything properly). I'm fixing some details and then I shall push :)

GivAlz commented:
Ok @luca-heltai , I updated and fixed everything and the test works.

There is one "important" change I decided to make: the triangulation was used only to extract the communicator, so it wasn't really useful as a triangulation.
I think it is better for the function to require the communicator as input and a vector of bounding boxes (and this is how it is implemented now).

In this way it can be used also for any process that needs to build a global tree out of some local bounding boxes.

GivAlz commented:
@luca-heltai , this PR has been sitting here for a while, but I don't understand what else I need to do/add/fix...

luca-heltai commented:
It seems that this cannot be compiled on Jenkins. Please check the logs there.

luca-heltai commented:
/rebuild

masterleinad commented:
/rebuild

masterleinad commented:
You need the following patch for this to compile without `MPI` support:
```
diff --git a/source/grid/grid_tools.cc b/source/grid/grid_tools.cc
index 0d30f87043..2da178fa23 100644
--- a/source/grid/grid_tools.cc
+++ b/source/grid/grid_tools.cc
@@ -5093,7 +5093,7 @@ namespace GridTools
     Assert(false,
            ExcMessage(
              "GridTools::build_global_description_tree() requires MPI."));
-    return {};
+    return RTree<std::pair<BoundingBox<spacedim>, unsigned int>>{};
 #else
     // Exchanging local bounding boxes
     std::vector<std::vector<BoundingBox<spacedim>>> global_bboxes;
```
since the constructor implicitly used is explicit.

GivAlz commented:
Thank you @masterleinad for catching that problem and solving it! I just added your patch :)

bangerth commented on source/grid/grid_tools.cc:
```diff
@@ -5081,6 +5081,65 @@ namespace GridTools
+    Assert(false,
+           ExcMessage(
+             "GridTools::build_global_description_tree() requires MPI."));
+    return RTree<std::pair<BoundingBox<spacedim>, unsigned int>>{};
```
I know I'm coming late to the party, but would it be easy to create an RTree from just the local description here? This would make it easier to run programs with and without MPI...

GivAlz replied:
Mmm...so you say it would be better to return a "one-box-tree" for running reasons? That's ok! I can change that :)

bangerth commented:
@GivAlz -- I've left another comment above. Also, there is no test in this pull request. Is this function tested elsewhere?

GivAlz commented:
@bangerth as for the test...my bad...it's sitting on my hard drive and never tracked, I shall add a PR with the test and, possibly, the return one-box tree (if that's what you were suggesting)
